addons/t9n/models/resource.py

```diff
@@ -14,3 +19,84 @@ class Resource(models.Model):
     project_id = fields.Many2one(
         comodel_name="t9n.project",
     )
+
+    _sql_constraints = [
+        (
+            "file_name_project_id_unique",
+            "unique(file_name, project_id)",
+            "A file with the same name already exists in the same project!"
+        ),
+    ]
+
+    def _decode_resource_file(self, resource_file):
+        try:
+            file_content = base64.b64decode(resource_file).decode()
+            po_obj = polib.pofile(file_content)
+        except (IOError, UnicodeDecodeError):
+            return None
+        return po_obj
+
+    def _get_message_from_file(self, resource_file):
+        for entry in resource_file:
+            references_list = [reference[0] + (f":{reference[1]}" if reference[1] else "")
+                               for reference in entry.occurrences]
```

Comment:
It's clearer if you unpack the values and give them a proper name:
```suggestion
            references_list = [fpath + (f":{lineno}" if lineno else "")
                               for fpath, lineno in entry.occurrences]
```

Reply:
Why do the "\n".join after instead of right here?